File: algorithm/sort/merge/merge.go
// go run merge.go
package main

import "fmt"

func main()  {
	arr := mergeSort([]int{1, 3, 2, 4, 5, 3})
	fmt.Printf("merge sort result %v", arr)
}

func mergeSort(arr []int) []int {
	length := len(arr)
	if length < 2 {
		return arr
	}

	middle := length / 2		
	left := arr[0:middle]
	right := arr[middle:]

	return merge(mergeSort(left), mergeSort(right))
}


func merge (left []int, right []int) []int {
	var result []int

	for len(left) != 0 && len(right) != 0 {
		if left[0] <= right[0] {
			   result = append(result, left[0])
			   left = left[1:]
		} else {
			   result = append(result, right[0])
			   right = right[1:]
		}
  	}
	
	for len(left) != 0 {
		result = append(result, left[0])
		left = left[1:]
 	}

  	for len(right) != 0 {
		result = append(result, right[0])
		right = right[1:]
	}

	return result
}

File: algorithm/sort/merge/merge.js
// node merge.js
/**
 * merge sort
 * @param {Array} arr
 * @return {Array}
 */
function mergeSort(arr) {
    var len = arr.length;
    if (len < 2) {
        return arr;
    }

    var middle = Math.floor(len / 2),
        left = arr.slice(0, middle),
        right = arr.slice(middle);

    return merge(mergeSort(left), mergeSort(right))
}

/**
 * merge arr
 * @param {Array} left
 * @param {Array} right
 * @return {Array}
 */
function merge(left, right) {
    var result = [];

    while (left.length && right.length) {
        if (left[0] > right[0]) {
            result.push(right.shift());
        } else {
            result.push(left.shift());
        }
    }

    while (right.length) {
        result.push(right.shift());
    }

    while (left.length) {
        result.push(left.shift());
    }

    return result;
}

console.log(mergeSort([1, 3, 4, 0, 7]));